File: 剑指offer/go/06.从尾到头打印链表.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

func (this *ListNode) add(x int) {
	var p = this
	for p.Next != nil {
		p = p.Next
	}
	p.Next = &ListNode{
		Val:  x,
		Next: nil,
	}
}

/*
 * 递归思路解决问题
 */
// func reversePrint(head *ListNode) []int {
// 	if head == nil {
// 		return []int{}
// 	}
// 	return append(reversePrint(head.Next), head.Val)
// }

/*
 * 非递归思路
 */
func reversePrint(head *ListNode) []int {
	tmp := []int{}
	for head != nil {
		tmp = append(tmp, head.Val)
		head = head.Next
	}
	for i := 0; i < len(tmp)/2; i++ {
		tmp[i], tmp[len(tmp)-1-i] = tmp[len(tmp)-1-i], tmp[i]
	}
	return tmp
}

func main() {
	head := &ListNode{
		Val:  1,
		Next: nil,
	}
	head.add(2)
	head.add(3)

	fmt.Print(reversePrint(head))

}

File: 剑指offer/java/06.从尾到头打印链表.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;
import java.util.Stack;


class Solution {
    public int[] reversePrint(ListNode head) {
        Stack<Integer> stack = new Stack<Integer>();

        List<Integer> a = new ArrayList<>();

        Queue<Integer> q = new LinkedList<>();
        while (head != null) {
            stack.add(head.val);
            head = head.next;
        }
        int[] ans = new int[stack.size()];
        // System.out.println(Arrays.toString(stack.toArray()));
        int size = stack.size();
        for (int i = 0; i < size; i ++) {
            ans[i] = stack.pop();
        }
        return ans;
    }

    public static void main(String[] args) {
        ListNode head = new ListNode(2);
        ListNode p = head;
        p.next = new ListNode(3);
        p = p.next;
        p.next = new ListNode(1);

        Solution solution = new Solution();
        
        System.out.println(Arrays.toString(solution.reversePrint(head)));
        
    }
}

class ListNode {
    int val;
    ListNode next;

    ListNode(int x) {
        val = x;
    }
}